### PR TITLE
Added four-fold star + drawMultipleWithRotationPoint

### PR DESCRIPTION
**Pull request by muhammad-abbady**
Four-fold star is pretty self-explanatory. drawMultipleWithRotationPoint allows multiple callbacks to be executed in order with each rotation pass.

### CODE REVIEW

File: src/designs/implementations/four-fold-star.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      center,
+      2,
+      90,
+      [topCenterPoint, rotatePoint(topCenterPoint, center, 180)],
+      ([rFrom, rTo]) => this.drawLine(rFrom, rTo, pencilBrush),
+    );
+
+    const circle = await this.drawCircle(center, radius, pencilBrush);
+
+    const diagonalsIntersectionsWithCircle = [];
```
Comment (OmarTawfik) [inline src/designs/implementations/four-fold-star.ts]: you can declare `diagonalsIntersectionsWithCircle` as:
```
const diagonalsIntersectionsWithCircle = [
  await (),
  await (),
];
````

File: src/designs/implementations/four-fold-star.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    diagonalsIntersectionsWithCircle.push(
+      intersectionBetweenLineAndCircle(diagonals[0].vertices[0], diagonals[0].vertices[1], center, radius),
+      intersectionBetweenLineAndCircle(diagonals[1].vertices[0], diagonals[1].vertices[1], center, radius),
+    );
+
+    const star = await this.drawMultipleWithRotatingPoint(
```
Comment (OmarTawfik) [inline src/designs/implementations/four-fold-star.ts]: I see why you added `drawMultipleWithRotatingPoint`. The original one doesn't enforce a certain type to be returned. so you can return an array (of two points) using the original method. In that case `star` will be a 2d array.

File: src/designs/implementations/four-fold-star.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      ],
+      ([rFrom, rTo]) => this.drawLine(rFrom, rTo, mainBorderBrush),
+      ([, , rFrom, rTo]) => this.drawLine(rFrom, rTo, mainBorderBrush),
+      ([rFrom, , , , rTo]) => this.drawLine(rFrom, rTo, mainBorderBrush),
+      ([, , , , , rFrom, rTo]) => this.drawLine(rFrom, rTo, mainBorderBrush),
+    );
+
+    this.removeAndUpdate(...square, circle, ...diagonals, ...axis);
+
+    for (const i of star) {
```
Comment (OmarTawfik) [inline src/designs/implementations/four-fold-star.ts]: I suggest renaming to `starGroups` then and using `for (const group of starGroups)`

File: src/designs/implementations/four-fold-star.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    super(scene, speed, token, shouldAnimate);
+  }
+
+  public async render(): Promise<void> {
+    const center = this.calculateCenterPoint();
+    const radius = Math.min(center.x, center.y);
+    const topCenterPoint = new Two.Vector(center.x, 0);
+    const topLeftCorner = new Two.Vector(0, 0);
+
+    const square = await this.drawWithRotatingPoint(
```
[the remark this answers is not in the document]
Comment (muhammad-abbady) [reply src/designs/implementations/four-fold-star.ts]: Agreed.

File: src/designs/implementations/four-fold-star.ts (hunk 1 above)
Comment (muhammad-abbady) [reply src/designs/implementations/four-fold-star.ts]: Done.

File: src/designs/implementations/four-fold-star.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      center,
+      4,
+      90,
+      [topLeftCorner, new Two.Vector(this.scene.width, 0)],
+      ([rFrom, rTo]) => this.drawLine(rFrom, rTo, pencilBrush),
+    );
+
+    const diagonals = await this.drawWithRotatingPoint(
```
[the remark this answers is not in the document]
Comment (muhammad-abbady) [reply src/designs/implementations/four-fold-star.ts]: Done.

File: src/designs/implementations/four-fold-star.ts (hunk 3 above)
Comment (muhammad-abbady) [reply src/designs/implementations/four-fold-star.ts]: Done.

File: src/designs/implementations/four-fold-star.ts (hunk 2 above)
Comment (muhammad-abbady) [reply src/designs/implementations/four-fold-star.ts]: Agreed. Done.

**Pull request merged**